src/mmap_xen.rs

```diff
@@ -0,0 +1,575 @@
+
+use bitflags::bitflags;
+use libc::{c_int, c_void, off_t, MAP_SHARED, PROT_READ, PROT_WRITE, _SC_PAGESIZE};
+use std::{
+    alloc::{self, handle_alloc_error, Layout},
+    io,
+    mem::{align_of, size_of},
+    os::raw::c_ulong,
+    os::unix::io::AsRawFd,
+    ptr::null_mut,
+    slice,
+    sync::{Arc, Mutex},
+};
+
+use vmm_sys_util::ioctl::{ioctl_expr, ioctl_with_ref, _IOC_NONE};
+
+use crate::guest_memory::GuestAddress;
+use crate::mmap::{MappedAddress, MappedAddressSimple, MmapInternal};
+use crate::mmap_unix::{self, Error, Result};
+
+fn page_size() -> u64 {
+    // SAFETY: Safe because this call just returns the page size and doesn't have any side effects.
+    unsafe { libc::sysconf(_SC_PAGESIZE) as u64 }
+}
+
+fn pages(size: usize) -> (usize, usize) {
+    let page_size = page_size() as usize;
+    let num = (size + page_size - 1) / page_size;
+
+    (num, page_size * num)
+}
+
+// Bit mask for the vhost-user xen mmap message.
+bitflags! {
+    /// Flags for the Xen mmap message.
+    pub struct MmapXenFlags: u32 {
+        /// Xen foreign memory (accessed via /dev/privcmd).
+        const FOREIGN = 0x1;
+        /// Xen grant memory (accessed via /dev/gntdev).
+        const GRANT = 0x2;
+        /// Xen no advance mapping.
+        const NO_ADVANCE_MAP = 0x8;
+        /// All valid mappings.
+        const ALL = Self::FOREIGN.bits() | Self::GRANT.bits();
+    }
+}
+
+impl MmapXenFlags {
+    /// Mmap flags are valid.
+    pub fn is_valid(flags: u32) -> bool {
+        // Only of one of FOREIGN or GRANT should be set and mmap_in_advance() should be true with
+        // FOREIGN.
+        if Self::is_foreign(flags) {
+            if Self::is_grant(flags) || !Self::mmap_in_advance(flags) {
+                return false;
+            }
+        } else if !Self::is_grant(flags) {
+            return false;
+        }
+
+        true
+    }
+
+    /// Is xen foreign memory.
+    pub fn is_foreign(flags: u32) -> bool {
+        flags & Self::FOREIGN.bits() != 0
+    }
+
+    /// Is xen grant memory.
+    pub fn is_grant(flags: u32) -> bool {
+        flags & Self::GRANT.bits() != 0
+    }
+
+    /// Can mmap entire region in advance.
+    pub fn mmap_in_advance(flags: u32) -> bool {
+        flags & Self::NO_ADVANCE_MAP.bits() == 0
+    }
+}
+
+/// Xen Foreign memory mapping interface.
+
+#[repr(C)]
+#[derive(Debug, Copy, Clone)]
+/// Privcmd mmap batch v2 command
+///
+/// include/uapi/xen/privcmd.h: `privcmd_mmapbatch_v2`
+struct PrivCmdMmapBatchV2 {
+    /// number of pages to populate
+    num: u32,
+    /// target domain
+    domid: u16,
+    /// virtual address
+    addr: *mut c_void,
+    /// array of mfns
+    arr: *const u64,
+    /// array of error codes
+    err: *mut c_int,
+}
+
+const XEN_PRIVCMD_TYPE: u32 = 'P' as u32;
+
+// #define IOCTL_PRIVCMD_MMAPBATCH_V2 _IOC(_IOC_NONE, 'P', 4, sizeof(privcmd_mmapbatch_v2_t))
+fn ioctl_privcmd_mmapbatch_v2() -> c_ulong {
```

Comment:
This should be `libc::Ioctl` I think? musl is `c_int`, right? But my feeling is that other libs are wrong about this too.

Reply:
That's what gets called internally here and it takes care of different types for libc, musl, etc.